### ISSUE TITLE
Issue using tars.predict in UDF

### ISSUE THREAD

**Issue by xfgavin** [flairNLP/flair#2740]
I got very different result when using tars.predict in a function in python with an udf wrapping the function and using it in spark sql. Here is the code:

```
pip install flair
from flair.data import Sentence
from flair.models import SequenceTagger
from pyspark.sql.functions import *
from pyspark.sql.types import *
from flair.datasets import GO_EMOTIONS
from flair.models import TARSClassifier
from flair.trainers import ModelTrainer
tars = TARSClassifier.load('tars-base')
tars.switch_to_task("GO_EMOTIONS")

def getEmotion(text):

    sentence = Sentence(text)

    tars.predict(sentence)

    result = sentence.to_dict()

    return result

spark.udf.register("getEmotionUDF", getEmotion,StringType())
```

In spark sql via databricks:
```
%sql

select getEmotionUDF('i am happy');
```

I got:

`
{all labels=[{value=NEUTRAL, confidence=0.7152036428451538}, {value=ANGER, confidence=0.7273235321044922}, {value=FEAR, confidence=0.6826728582382202}, {value=ANNOYANCE, confidence=0.7781732678413391}, {value=SURPRISE, confidence=0.6873858571052551}, {value=GRATITUDE, confidence=0.7292483448982239}, {value=DESIRE, confidence=0.7011390328407288}, {value=OPTIMISM, confidence=0.7458798885345459}, {value=ADMIRATION, confidence=0.7463255524635315}, {value=CONFUSION, confidence=0.7086165547370911}, {value=AMUSEMENT, confidence=0.676873505115509}, {value=APPROVAL, confidence=0.7240709662437439}, {value=CARING, confidence=0.7137355208396912}, {value=EMBARRASSMENT, confidence=0.6914198398590088}, {value=REALIZATION, confidence=0.7595157027244568}, {value=DISAPPOINTMENT, confidence=0.7149665355682373}, {value=GRIEF, confidence=0.7443049550056458}, {value=SADNESS, confidence=0.7657970786094666}, {value=CURIOSITY, confidence=0.75248783826828}, {value=JOY, confidence=0.7151253819465637}, {value=LOVE, confidence=0.7284602522850037}, {value=EXCITEMENT, confidence=0.7087971568107605}, {value=DISAPPROVAL, confidence=0.7449938058853149}, {value=REMORSE, confidence=0.6994790434837341}, {value=DISGUST, confidence=0.7635707259178162}, {value=RELIEF, confidence=0.760797381401062}, {value=PRIDE, confidence=0.7652835249900818}, {value=NERVOUSNESS, confidence=0.7355543971061707}], text=i am happy}
`

When using the function directly in python as below:
```
getEmotion("I am happy")
```
I got:
`
{'text': 'I am happy', 'all labels': [{'value': 'JOY', 'confidence': 0.9816395044326782}]}
`

**Comment by xfgavin** [comment]
@kashif @pranaychandekar @apohllo @mhoyer @heukirne @ thought on this issue?
Thanks.

**Comment by tulha** [comment]
Facing a similar issue. Would be great if you could take a look!

**Comment by aabouelleil-seniorlink** [comment]
I am having a very similar problem as well, and have tried multiple solutions that haven't worked, but I'll share them here.

### The Setup

AWS EMR cluster with several worker nodes. Flair 0.11 was installed via bootstrap script. Code was run in an EMR Studio Notebook. 

### Code

First the import and initialization of the classifier
```
from flair.models import TextClassifier
from flair.data import Sentence
#Locally run 
classifier = TextClassifier.load('en-sentiment')
```
This resulted in a logging error as follows, but I continued processing the next bit of code anyway. 
```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.7/logging/__init__.py", line 1028, in emit
    stream.write(msg + self.terminator)
  File "/tmp/9205348622636130180", line 534, in write
    super(UnicodeDecodingStringIO, self).write(s)
ValueError: I/O operation on closed file
Call stack:
  File "/tmp/9205348622636130180", line 714, in <module>
    sys.exit(main())
  File "/tmp/9205348622636130180", line 686, in main
    response = handler(content)
  File "/tmp/9205348622636130180", line 318, in execute_request
    result = node.execute()
  File "/tmp/9205348622636130180", line 229, in execute
    exec(code, global_dict)
  File "<stdin>", line 4, in <module>
  File "/usr/local/lib/python3.7/site-packages/flair/nn/model.py", line 141, in load
    f = file_utils.load_big_file(str(model_file))
  File "/usr/local/lib/python3.7/site-packages/flair/file_utils.py", line 33, in load_big_file
    logger.info(f"loading file {f}")
Message: 'loading file /var/lib/livy/.flair/models/sentiment-en-mix-distillbert_4.pt'
Arguments: ()
```

I then execute the sentiment analysis as follows, which returns a result that seems reasonable given the sentence:
```
from flair.models import TextClassifier
from flair.data import Sentence
import pyspark.sql.functions as F

def predict(text):
    sentence = Sentence(text)
    classifier.predict(sentence)
    results = sentence.labels[0].to_dict()
    return [results['value'], results['confidence']]

predict_udf = F.udf(predict, ArrayType(StringType()))
text = "While we are thankful for the praise, be SURE to give yourself a huge pat on the back as well. Caregiving is no easy feat, there will be mountains and valleys- but rest assured, you are so appreciated."
print(predict(text))
```

Resulting in:

`['POSITIVE', 0.9947469830513]`


Next, executing this same via a UDF (adding to the code above):
```
spark = SparkSession.builder.appName("Spark-Sentiment").getOrCreate()
data = [(text)]
df = spark.createDataFrame(data=data, schema = StringType())
df.withColumn('result', predict_udf('value')).select('result').show()
```

Resulting in:

+--------------------+
|              result|
+--------------------+
|[POSITIVE, 0.8322...|
+--------------------+

What's more disturbing, is the result changed with subsequent runs of the same cell in EMR Studio:
next run:
+--------------------+
|              result|
+--------------------+
|[NEGATIVE, 0.6094...|
+--------------------+

and next run:
+--------------------+
|              result|
+--------------------+
|[POSITIVE, 0.5158...|
+--------------------+

## Attempted Solutions

My first theory of this problem is that because Flair downloads the text classifier corpus, that it's downloading on the master node of my cluster but not the worker nodes. Since the UDF gets executed on a worker node, perhaps it's not accessing the classifier appropriately.

To test this, I tried making the classifier a Spark broadcast variable, assuming that the classifier would then be distributed to the worker nodes:

```
global_classifier = spark.sparkContext.broadcast(classifier)

def predict(text):
    sentence = Sentence(text)
    global_classifier.value.predict(sentence)
    results = sentence.labels[0].to_dict()
    return [results['value'], str(results['confidence'])]
```

I then ran:
`print(predict(text))`

And got:
`['POSITIVE', '0.7753053307533264']`

The result, for some reason, is different in terms of confidence score from when I ran it the first time.

Next, via UDF:

```
predict_udf = F.udf(predict, ArrayType(StringType()))
data = [(text)]
df = spark.createDataFrame(data=data, schema = StringType())
df.withColumn('result', predict_udf('value')).select('result').show()
```

Resulting in what seems to be again, non-deterministic results:
run 1
+--------------------+
|              result|
+--------------------+
|[POSITIVE, 0.9970...|
+--------------------+
run 2
+--------------------+
|              result|
+--------------------+
|[NEGATIVE, 0.7868...|
+--------------------+


To summarize, the problems seem to be, Flair run 'locally' seems to give consistent and reasonable results, but when run via a UDF and distributed to a node gives conflicting and non-deterministic results. 

It's not clear to me if this is an issue with Flair or with my EMR environment, but some input from developers would help in figuring this out.
